File: tests/requests/ai_spec.rs
use loco_rs::testing::prelude::*;
use serial_test::serial;
use sha2::{Digest, Sha256};
use wiremock::matchers::{header, method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};

use bookkeeping_backend::app::App;

use super::prepare_data::{create_import_log, find_log, jpeg_bytes, json_body, register_and_login};

fn configure(server: &MockServer) {
    std::env::set_var("DEEPSEEK_BASE_URL", server.uri());
    std::env::set_var("DEEPSEEK_API_KEY", "test-key");
    std::env::set_var("DEEPSEEK_MODEL", "deepseek-flash");
    std::env::set_var("LIHKG_ALLOWED_HOSTS", "127.0.0.1");
    std::env::set_var("LIHKG_UPLOAD_URL", format!("{}/upload", server.uri()));
}

fn sha256_hex(bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    hex::encode(hasher.finalize())
}

fn deepseek_body(parsed: &serde_json::Value) -> serde_json::Value {
    serde_json::json!({
        "choices": [{ "message": { "content": parsed.to_string() } }],
        "usage": { "prompt_tokens": 10, "completion_tokens": 8 }
    })
}

fn multipart_body(filename: &str, content_type: &str, bytes: &[u8]) -> Vec<u8> {
    let boundary = "----bookkeeping-test-boundary";
    let mut body = Vec::new();
    body.extend_from_slice(format!("--{boundary}\r\n").as_bytes());
    body.extend_from_slice(
        format!("Content-Disposition: form-data; name=\"file\"; filename=\"{filename}\"\r\n")
            .as_bytes(),
    );
    body.extend_from_slice(format!("Content-Type: {content_type}\r\n\r\n").as_bytes());
    body.extend_from_slice(bytes);
    body.extend_from_slice(format!("\r\n--{boundary}--\r\n").as_bytes());
    body
}

#[tokio::test]
#[serial]
async fn upload_sends_lihkg_origin_and_returns_digest() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        Mock::given(method("POST"))
            .and(path("/upload"))
            .and(header("origin", "https://lihkg.com"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "url": "https://img.eservice-hk.net/a.jpg"
            })))
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let jpeg = jpeg_bytes();
        let body = multipart_body("receipt.jpg", "image/jpeg", &jpeg);
        let response = request
            .post("/api/v1/receipts/upload")
            .add_header(name, value)
            .content_type("multipart/form-data; boundary=----bookkeeping-test-boundary")
            .bytes(body.into())
            .await;
        assert_eq!(response.status_code(), 201);
        let response_body = json_body(&response.text());
        assert!(response_body["data"]["url"]
            .as_str()
            .unwrap()
            .ends_with("a.jpg"));
        assert_eq!(response_body["data"]["sha256"].as_str().unwrap().len(), 64);
        assert_eq!(response_body["data"]["sha256"], sha256_hex(&jpeg));

        let requests = server.received_requests().await.unwrap();
        let upload = requests
            .iter()
            .find(|request| request.url.path() == "/upload")
            .expect("upload request");
        assert_eq!(
            upload.headers.get("origin").unwrap().to_str().unwrap(),
            "https://lihkg.com"
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn upload_rejects_non_image_file() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let body = multipart_body("receipt.exe", "application/octet-stream", b"MZxxxx");
        let response = request
            .post("/api/v1/receipts/upload")
            .add_header(name, value)
            .content_type("multipart/form-data; boundary=----bookkeeping-test-boundary")
            .bytes(body.into())
            .await;
        assert_eq!(response.status_code(), 422);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn parse_caches_and_rejects_non_whitelisted_hosts() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let jpeg = jpeg_bytes();
        Mock::given(method("GET"))
            .and(path("/receipt.jpg"))
            .respond_with(ResponseTemplate::new(200).set_body_raw(jpeg.clone(), "image/jpeg"))
            .mount(&server)
            .await;
        Mock::given(method("POST"))
            .and(path("/chat/completions"))
            .respond_with(ResponseTemplate::new(200).set_body_json(deepseek_body(
                &serde_json::json!({
                    "amount_cents": 1234,
                    "kind": "expense",
                    "occurred_at": "2026-09-14T10:00:00+08:00",
                    "confidence": 0.9
                }),
            )))
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let image_url = format!("{}/receipt.jpg", server.uri());

        for _ in 0..2 {
            let response = request
                .post("/api/v1/ai/parse")
                .add_header(name.clone(), value.clone())
                .json(&serde_json::json!({ "image_url": image_url }))
                .await;
            assert_eq!(response.status_code(), 200);
        }

        let deepseek_calls = server
            .received_requests()
            .await
            .unwrap()
            .iter()
            .filter(|request| request.url.path() == "/chat/completions")
            .count();
        assert_eq!(deepseek_calls, 1, "second parse should hit the cache");

        let ssrf = request
            .post("/api/v1/ai/parse")
            .add_header(name, value)
            .json(&serde_json::json!({ "image_url": "http://169.254.169.254/private" }))
            .await;
        assert_eq!(ssrf.status_code(), 400);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn parse_ignores_a_stale_cache_entry() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let jpeg = jpeg_bytes();
        Mock::given(method("GET"))
            .and(path("/stale.jpg"))
            .respond_with(ResponseTemplate::new(200).set_body_raw(jpeg.clone(), "image/jpeg"))
            .mount(&server)
            .await;
        Mock::given(method("POST"))
            .and(path("/chat/completions"))
            .respond_with(ResponseTemplate::new(200).set_body_json(deepseek_body(
                &serde_json::json!({
                    "amount_cents": 4500,
                    "kind": "expense",
                    "occurred_at": "2026-02-21T15:45:00",
                    "category_hint": "飲食",
                    "confidence": 0.9
                }),
            )))
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let stale = create_import_log(
            &ctx.db,
            &fixture.user_id,
            &sha256_hex(&jpeg),
            serde_json::json!({ "amount_cents": 1, "kind": "expense", "category_hint": null }),
        )
        .await;

        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let response = request
            .post("/api/v1/ai/parse")
            .add_header(name, value)
            .json(&serde_json::json!({ "image_url": format!("{}/stale.jpg", server.uri()) }))
            .await;
        assert_eq!(response.status_code(), 200);
        let body = json_body(&response.text());
        assert_ne!(body["data"]["id"], stale.id);
        assert_eq!(body["data"]["parsed"]["amount_cents"], 4500);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn parse_handles_multibyte_and_prefix_markers_and_hint_normalization() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let jpeg = jpeg_bytes();
        for name in [
            "/cn.jpg",
            "/prefix.jpg",
            "/case.jpg",
            "/nomatch.jpg",
            "/kind.jpg",
        ] {
            Mock::given(method("GET"))
                .and(path(name))
                .respond_with(ResponseTemplate::new(200).set_body_raw(jpeg.clone(), "image/jpeg"))
                .mount(&server)
                .await;
        }
        Mock::given(method("POST"))
            .and(path("/chat/completions"))
            .respond_with(ResponseTemplate::new(200).set_body_json(deepseek_body(
                &serde_json::json!({
                    "amount_cents": 1234,
                    "kind": "expense",
                    "occurred_at": "2026-09-14T10:00:00+08:00",
                    "merchant_name": "茶餐廳",
                    "note": "午餐",
                    "category_hint": "飲食",
                    "confidence": 0.9
                }),
            )))
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let response = request
            .post("/api/v1/ai/parse")
            .add_header(name, value)
            .json(&serde_json::json!({ "image_url": format!("{}/cn.jpg", server.uri()) }))
            .await;
        assert_eq!(response.status_code(), 200);
        assert_eq!(
            json_body(&response.text())["data"]["parsed"]["merchant_name"],
            "茶餐廳"
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn parse_matches_category_hint_against_user_categories() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let jpeg = jpeg_bytes();
        Mock::given(method("GET"))
            .and(path("/cat.jpg"))
            .respond_with(ResponseTemplate::new(200).set_body_raw(jpeg.clone(), "image/jpeg"))
            .mount(&server)
            .await;
        Mock::given(method("POST"))
            .and(path("/chat/completions"))
            .respond_with(ResponseTemplate::new(200).set_body_json(deepseek_body(
                &serde_json::json!({
                    "amount_cents": 1234,
                    "kind": "expense",
                    "occurred_at": "2026-09-14T10:00:00+08:00",
                    "category_hint": "飲食",
                    "confidence": 0.9
                }),
            )))
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let response = request
            .post("/api/v1/ai/parse")
            .add_header(name, value)
            .json(&serde_json::json!({ "image_url": format!("{}/cat.jpg", server.uri()) }))
            .await;
        let body = json_body(&response.text());
        assert_eq!(body["data"]["parsed"]["category_hint"], "飲食");
        assert_eq!(
            body["data"]["suggested_category_id"],
            fixture.expense_category_id
        );

        // The prompt must include the user's expense categories.
        let requests = server.received_requests().await.unwrap();
        let deepseek = requests
            .iter()
            .find(|request| request.url.path() == "/chat/completions")
            .unwrap();
        let request_body: serde_json::Value = serde_json::from_slice(&deepseek.body).unwrap();
        let prompt = request_body["messages"][0]["content"][0]["text"]
            .as_str()
            .unwrap();
        assert!(prompt.contains("EXPENSE categories: [\"飲食\""));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn parse_forces_unmatched_or_wrong_kind_hints_to_null() {
    request::<App, _, _>(|request, ctx| async move {
        let server = MockServer::start().await;
        configure(&server);
        let jpeg = jpeg_bytes();
        for name in ["/nomatch.jpg", "/kind.jpg", "/case.jpg"] {
            Mock::given(method("GET"))
                .and(path(name))
                .respond_with(ResponseTemplate::new(200).set_body_raw(jpeg.clone(), "image/jpeg"))
                .mount(&server)
                .await;
        }
        Mock::given(method("POST"))
            .and(path("/chat/completions"))
            .respond_with(|request: &wiremock::Request| {
                let body: serde_json::Value = serde_json::from_slice(&request.body).unwrap();
                let prompt = body["messages"][0]["content"][0]["text"].as_str().unwrap();
                if prompt.contains("Groceries") {
                    ResponseTemplate::new(200).set_body_json(deepseek_body(&serde_json::json!({
                        "amount_cents": 1234,
                        "kind": "expense",
                        "occurred_at": "2026-09-14T10:00:00+08:00",
                        "category_hint": "groceries",
                        "confidence": 0.9
                    })))
                } else {
                    ResponseTemplate::new(200).set_body_json(deepseek_body(&serde_json::json!({
                        "amount_cents": 1234,
                        "kind": "expense",
                        "occurred_at": "2026-09-14T10:00:00+08:00",
                        "category_hint": "餐飲",
                        "confidence": 0.9
                    })))
                }
            })
            .mount(&server)
            .await;

        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = super::prepare_data::auth_header(&fixture.token);

        let unmatched = request
            .post("/api/v1/ai/parse")
            .add_header(name.clone(), value.clone())
            .json(&serde_json::json!({ "image_url": format!("{}/nomatch.jpg", server.uri()) }))
            .await;
        let body = json_body(&unmatched.text());
        assert_eq!(
            body["data"]["parsed"]["category_hint"],
            serde_json::Value::Null
        );
        assert_eq!(
            body["data"]["suggested_category_id"],
            serde_json::Value::Null
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn confirm_creates_an_ai_transaction() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let log = create_import_log(
            &ctx.db,
            &fixture.user_id,
            &"a".repeat(64),
            serde_json::json!({
                "amount_cents": 500,
                "kind": "expense",
                "occurred_at": "2026-09-14T10:00:00+08:00"
            }),
        )
        .await;

        let (name, value) = super::prepare_data::auth_header(&fixture.token);
        let response = request
            .post("/api/v1/ai/confirm")
            .add_header(name, value)
            .json(&serde_json::json!({
                "ai_import_log_id": log.id,
                "account_id": fixture.account_id,
                "amount_cents": 500,
                "kind": "expense",
                "occurred_at": "2026-09-14T10:00:00+08:00"
            }))
            .await;
        assert_eq!(response.status_code(), 201);
        let body = json_body(&response.text());
        assert_eq!(body["data"]["source"], "ai");
        assert_eq!(
            body["data"]["image_urls"][0],
            "https://img.eservice-hk.net/a.jpg"
        );
        assert_eq!(
            find_log(&ctx.db, &log.id).await.transaction_id.unwrap(),
            body["data"]["id"].as_str().unwrap()
        );
    })
    .await;
}

File: test/ai.spec.ts
import {INestApplication} from "@nestjs/common";
import {afterAll, afterEach, beforeAll, beforeEach, describe, expect, it} from "vitest";

import {sha256Hex} from "../src/common/util";
import {DataSource} from "typeorm";
import {createApp, http, HttpClient, resetDatabase} from "./helpers/app";
import {authHeader, createImportLog, findLog, jpegBytes, registerAndLogin} from "./helpers/fixtures";
import {MockResponse, MockServer, startMockServer} from "./helpers/mock-server";

function deepseekBody(parsed: unknown): string {
    return JSON.stringify({
        choices: [{message: {content: JSON.stringify(parsed)}}],
        usage: {prompt_tokens: 10, completion_tokens: 8},
    });
}

function multipartBody(filename: string, contentType: string, bytes: Buffer): Buffer {
    const boundary = "----bookkeeping-test-boundary";
    return Buffer.concat([
        Buffer.from(`--${boundary}\r\n`),
        Buffer.from(`Content-Disposition: form-data; name="file"; filename="${filename}"\r\n`),
        Buffer.from(`Content-Type: ${contentType}\r\n\r\n`),
        bytes,
        Buffer.from(`\r\n--${boundary}--\r\n`),
    ]);
}

describe("receipts & AI", () => {
    let app: INestApplication;
    let client: HttpClient;
    let dataSource: DataSource;
    const servers: MockServer[] = [];

    beforeAll(async () => {
        app = await createApp();
        client = http(app);
        dataSource = app.get(DataSource);
    });

    beforeEach(async () => {
        await resetDatabase(dataSource);
    });

    afterEach(async () => {
        await Promise.all(servers.splice(0).map(server => server.close()));
    });

    afterAll(async () => {
        await app.close();
    });

    async function mockServer(handler: (request: Parameters<Parameters<typeof startMockServer>[0]>[0]) => MockResponse) {
        const server = await startMockServer(handler);
        servers.push(server);
        process.env.DEEPSEEK_BASE_URL = server.uri;
        process.env.DEEPSEEK_API_KEY = "test-key";
        process.env.DEEPSEEK_MODEL = "deepseek-flash";
        process.env.LIHKG_ALLOWED_HOSTS = "127.0.0.1";
        process.env.LIHKG_UPLOAD_URL = `${server.uri}/upload`;
        return server;
    }

    it("upload sends lihkg origin and returns digest", async () => {
        const server = await mockServer(request =>
            request.path === "/upload" && request.method === "POST"
                ? {
                      status: 200,
                      headers: {"content-type": "application/json"},
                      body: JSON.stringify({url: "https://img.eservice-hk.net/a.jpg"}),
                  }
                : {status: 404}
        );

        const fixture = await registerAndLogin(client);
        const jpeg = jpegBytes();
        const body = multipartBody("receipt.jpg", "image/jpeg", jpeg);
        const response = await client.post("/api/v1/receipts/upload").set(authHeader(fixture.token)).set("Content-Type", "multipart/form-data; boundary=----bookkeeping-test-boundary").send(body);

        expect(response.status).toBe(201);
        expect(response.body.data.url).toMatch(/a\.jpg$/);
        expect(response.body.data.sha256).toHaveLength(64);
        expect(response.body.data.sha256).toBe(sha256Hex(jpeg));

        const upload = server.requests.find(request => request.path === "/upload");
        expect(upload?.headers.origin).toBe("https://lihkg.com");
    });

    it("upload rejects non-image file", async () => {
        await mockServer(() => ({status: 404}));
        const fixture = await registerAndLogin(client);
        const body = multipartBody("receipt.exe", "application/octet-stream", Buffer.from("MZxxxx"));

        const response = await client.post("/api/v1/receipts/upload").set(authHeader(fixture.token)).set("Content-Type", "multipart/form-data; boundary=----bookkeeping-test-boundary").send(body);
        expect(response.status).toBe(422);
    });

    it("parse caches and rejects non-whitelisted hosts", async () => {
        const jpeg = jpegBytes();
        const server = await mockServer(request => {
            if (request.method === "GET" && request.path === "/receipt.jpg") {
                return {status: 200, headers: {"content-type": "image/jpeg"}, body: jpeg};
            }
            if (request.method === "POST" && request.path === "/chat/completions") {
                return {
                    status: 200,
                    headers: {"content-type": "application/json"},
                    body: deepseekBody({
                        amount_cents: 1234,
                        kind: "expense",
                        occurred_at: "2026-09-14T10:00:00+08:00",
                        confidence: 0.9,
                    }),
                };
            }
            return {status: 404};
        });

        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);
        const imageUrl = `${server.uri}/receipt.jpg`;

        for (let index = 0; index < 2; index += 1) {
            const response = await client.post("/api/v1/ai/parse").set(headers).send({image_url: imageUrl});
            expect(response.status).toBe(200);
        }

        expect(server.countRequests("/chat/completions")).toBe(1);

        const ssrf = await client.post("/api/v1/ai/parse").set(headers).send({image_url: "http://169.254.169.254/private"});
        expect(ssrf.status).toBe(400);
    });

    it("parse ignores a stale cache entry", async () => {
        const jpeg = jpegBytes();
        const server = await mockServer(request => {
            if (request.method === "GET" && request.path === "/stale.jpg") {
                return {status: 200, headers: {"content-type": "image/jpeg"}, body: jpeg};
            }
            if (request.method === "POST" && request.path === "/chat/completions") {
                return {
                    status: 200,
                    headers: {"content-type": "application/json"},
                    body: deepseekBody({
                        amount_cents: 4500,
                        kind: "expense",
                        occurred_at: "2026-02-21T15:45:00",
                        category_hint: "飲食",
                        confidence: 0.9,
                    }),
                };
            }
            return {status: 404};
        });

        const fixture = await registerAndLogin(client);
        const stale = await createImportLog(dataSource, fixture.userId, sha256Hex(jpeg), {
            amount_cents: 1,
            kind: "expense",
            category_hint: null,
        });

        const response = await client
            .post("/api/v1/ai/parse")
            .set(authHeader(fixture.token))
            .send({image_url: `${server.uri}/stale.jpg`});
        expect(response.status).toBe(200);
        expect(response.body.data.id).not.toBe(stale.id);
        expect(response.body.data.parsed.amount_cents).toBe(4500);
    });

    it("parse handles multibyte text", async () => {
        const jpeg = jpegBytes();
        const server = await mockServer(request => {
            if (request.method === "GET" && request.path === "/cn.jpg") {
                return {status: 200, headers: {"content-type": "image/jpeg"}, body: jpeg};
            }
            if (request.method === "POST" && request.path === "/chat/completions") {
                return {
                    status: 200,
                    headers: {"content-type": "application/json"},
                    body: deepseekBody({
                        amount_cents: 1234,
                        kind: "expense",
                        occurred_at: "2026-09-14T10:00:00+08:00",
                        merchant_name: "茶餐廳",
                        note: "午餐",
                        category_hint: "飲食",
                        confidence: 0.9,
                    }),
                };
            }
            return {status: 404};
        });

        const fixture = await registerAndLogin(client);
        const response = await client
            .post("/api/v1/ai/parse")
            .set(authHeader(fixture.token))
            .send({image_url: `${server.uri}/cn.jpg`});
        expect(response.status).toBe(200);
        expect(response.body.data.parsed.merchant_name).toBe("茶餐廳");
    });

    it("parse matches category hint against user categories", async () => {
        const jpeg = jpegBytes();
        const server = await mockServer(request => {
            if (request.method === "GET" && request.path === "/cat.jpg") {
                return {status: 200, headers: {"content-type": "image/jpeg"}, body: jpeg};
            }
            if (request.method === "POST" && request.path === "/chat/completions") {
                return {
                    status: 200,
                    headers: {"content-type": "application/json"},
                    body: deepseekBody({
                        amount_cents: 1234,
                        kind: "expense",
                        occurred_at: "2026-09-14T10:00:00+08:00",
                        category_hint: "飲食",
                        confidence: 0.9,
                    }),
                };
            }
            return {status: 404};
        });

        const fixture = await registerAndLogin(client);
        const response = await client
            .post("/api/v1/ai/parse")
            .set(authHeader(fixture.token))
            .send({image_url: `${server.uri}/cat.jpg`});
        expect(response.body.data.parsed.category_hint).toBe("飲食");
        expect(response.body.data.suggested_category_id).toBe(fixture.expenseCategoryId);

        const deepseek = server.requests.find(request => request.path === "/chat/completions");
        const requestBody = JSON.parse(deepseek!.body.toString("utf8"));
        const prompt = requestBody.messages[0].content[0].text as string;
        expect(prompt).toContain('EXPENSE categories: ["飲食"');
    });

    it("parse forces unmatched or wrong kind hints to null", async () => {
        const jpeg = jpegBytes();
        const server = await mockServer(request => {
            if (request.method === "GET" && request.path === "/nomatch.jpg") {
                return {status: 200, headers: {"content-type": "image/jpeg"}, body: jpeg};
            }
            if (request.method === "POST" && request.path === "/chat/completions") {
                const body = JSON.parse(request.body.toString("utf8"));
                const prompt = body.messages[0].content[0].text as string;
                return {
                    status: 200,
                    headers: {"content-type": "application/json"},
                    body: deepseekBody({
                        amount_cents: 1234,
                        kind: "expense",
                        occurred_at: "2026-09-14T10:00:00+08:00",
                        category_hint: prompt.includes("Groceries") ? "groceries" : "餐飲",
                        confidence: 0.9,
                    }),
                };
            }
            return {status: 404};
        });

        const fixture = await registerAndLogin(client);
        const response = await client
            .post("/api/v1/ai/parse")
            .set(authHeader(fixture.token))
            .send({image_url: `${server.uri}/nomatch.jpg`});
        expect(response.body.data.parsed.category_hint).toBeNull();
        expect(response.body.data.suggested_category_id).toBeNull();
    });

    it("confirm creates an ai transaction", async () => {
        const fixture = await registerAndLogin(client);
        const log = await createImportLog(dataSource, fixture.userId, "a".repeat(64), {
            amount_cents: 500,
            kind: "expense",
            occurred_at: "2026-09-14T10:00:00+08:00",
        });

        const response = await client.post("/api/v1/ai/confirm").set(authHeader(fixture.token)).send({
            ai_import_log_id: log.id,
            account_id: fixture.accountId,
            amount_cents: 500,
            kind: "expense",
            occurred_at: "2026-09-14T10:00:00+08:00",
        });
        expect(response.status).toBe(201);
        expect(response.body.data.source).toBe("ai");
        expect(response.body.data.image_urls[0]).toBe("https://img.eservice-hk.net/a.jpg");

        const updated = await findLog(dataSource, log.id);
        expect(updated.transaction_id).toBe(response.body.data.id);
    });

    it("interpret parses a natural-language sentence into a text preview", async () => {
        const server = await mockServer(request =>
            request.method === "POST" && request.path === "/chat/completions"
                ? {
                      status: 200,
                      headers: {"content-type": "application/json"},
                      body: deepseekBody({transactions: [{amount_cents: 4500, kind: "expense", occurred_at: "2026-09-14T10:00:00+08:00", merchant_name: "茶餐廳", confidence: 0.9}]}),
                  }
                : {status: 404}
        );

        const fixture = await registerAndLogin(client);
        const response = await client.post("/api/v1/ai/interpret").set(authHeader(fixture.token)).send({text: "尋日茶餐廳 45 蚊"});

        expect(response.status).toBe(200);
        expect(response.body.data.source).toBe("text");
        expect(response.body.data.image_urls).toEqual([]);
        expect(response.body.data.parsed.amount_cents).toBe(4500);
        expect(response.body.data.parsed_items).toHaveLength(1);
        expect(response.body.data.parsed_items[0].parsed.amount_cents).toBe(4500);

        const cached = await client.post("/api/v1/ai/interpret").set(authHeader(fixture.token)).send({text: "尋日茶餐廳 45 蚊"});
        expect(cached.body.data.id).toBe(response.body.data.id);
        expect(cached.body.data.parsed_items).toHaveLength(1);
        expect(server.countRequests("/chat/completions")).toBe(1);
    });

    it("interpret splits a multi-transaction sentence into one item per transaction", async () => {
        await mockServer(request =>
            request.method === "POST" && request.path === "/chat/completions"
                ? {
                      status: 200,
                      headers: {"content-type": "application/json"},
                      body: deepseekBody({
                          transactions: [
                              {amount_cents: 3000, kind: "expense", occurred_at: "2026-09-14T08:00:00+08:00", merchant_name: "茶餐廳", confidence: 0.9},
                              {amount_cents: 5000, kind: "expense", occurred_at: "2026-09-14T12:00:00+08:00", confidence: 0.9},
                              {amount_cents: 2000, kind: "expense", occurred_at: "2026-09-14T18:00:00+08:00", confidence: 0.9},
                          ],
                      }),
                  }
                : {status: 404}
        );

        const fixture = await registerAndLogin(client);
        const response = await client.post("/api/v1/ai/interpret").set(authHeader(fixture.token)).send({text: "早餐 30 午餐 50 車費 20"});

        expect(response.status).toBe(200);
        expect(response.body.data.parsed.amount_cents).toBe(3000);
        expect(response.body.data.parsed_items.map((item: {parsed: {amount_cents: number}}) => item.parsed.amount_cents)).toEqual([3000, 5000, 2000]);
    });

    it("interpret rejects blank or oversized text", async () => {
        await mockServer(() => ({status: 404}));
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);

        const blank = await client.post("/api/v1/ai/interpret").set(headers).send({text: "   "});
        expect(blank.status).toBe(422);

        const tooLong = await client
            .post("/api/v1/ai/interpret")
            .set(headers)
            .send({text: "a".repeat(501)});
        expect(tooLong.status).toBe(422);
    });

    it("interpret maps a DeepSeek failure to a 502", async () => {
        await mockServer(request => (request.path === "/chat/completions" ? {status: 500} : {status: 404}));
        const fixture = await registerAndLogin(client);

        const response = await client.post("/api/v1/ai/interpret").set(authHeader(fixture.token)).send({text: "午餐 45 蚊"});
        expect(response.status).toBe(502);
    });

    it("confirm materialises a transaction from a natural-language preview", async () => {
        const server = await mockServer(request =>
            request.method === "POST" && request.path === "/chat/completions"
                ? {
                      status: 200,
                      headers: {"content-type": "application/json"},
                      body: deepseekBody({amount_cents: 4500, kind: "expense", occurred_at: "2026-09-14T10:00:00+08:00", confidence: 0.9}),
                  }
                : {status: 404}
        );

        const fixture = await registerAndLogin(client);
        const preview = await client.post("/api/v1/ai/interpret").set(authHeader(fixture.token)).send({text: "午餐 45 蚊"});
        expect(preview.status).toBe(200);

        const response = await client.post("/api/v1/ai/confirm").set(authHeader(fixture.token)).send({
            ai_import_log_id: preview.body.data.id,
            account_id: fixture.accountId,
            amount_cents: 4500,
            kind: "expense",
            occurred_at: "2026-09-14T10:00:00+08:00",
        });

        expect(response.status).toBe(201);
        expect(response.body.data.source).toBe("ai");
        expect(response.body.data.image_urls).toEqual([]);

        const stored = server.requests.find(request => request.path === "/chat/completions");
        expect(stored).toBeDefined();
    });
});
